homeassistant/bootstrap.py

```diff
@@ -890,64 +890,39 @@ async def _async_set_up_integrations(
+        *(
+            (name, domain_group & domains_to_setup, timeout)
+            for name, domain_group, timeout in STAGE_0_INTEGRATIONS
+        ),
+        ("stage 1", STAGE_1_INTEGRATIONS & domains_to_setup, STAGE_1_TIMEOUT),
     ]
 
-    # calculate what components to setup in what stage
-    stage_1_domains: set[str] = set()
-
-    # Find all dependencies of any dependency of any stage 1 integration that
-    # we plan on loading and promote them to stage 1. This is done only to not
-    # get misleading log messages
-    deps_promotion: set[str] = STAGE_1_INTEGRATIONS
-    while deps_promotion:
-        old_deps_promotion = deps_promotion
-        deps_promotion = set()
-
-        for domain in old_deps_promotion:
-            if domain not in domains_to_setup or domain in stage_1_domains:
-                continue
-
-            stage_1_domains.add(domain)
-
-            if (dep_itg := integration_cache.get(domain)) is None:
-                continue
-
-            deps_promotion.update(dep_itg.all_dependencies)
-
-    stage_2_domains = domains_to_setup - stage_1_domains
-
-    for name, domain_group in pre_stage_domains:
-        if domain_group:
-            stage_2_domains -= domain_group
-            _LOGGER.info("Setting up %s: %s", name, domain_group)
-            to_be_loaded = domain_group.copy()
-            to_be_loaded.update(
-                dep
-                for domain in domain_group
-                if (integration := integration_cache.get(domain)) is not None
-                for dep in integration.all_dependencies
-            )
-            async_set_domains_to_be_loaded(hass, to_be_loaded)
-            await _async_setup_multi_components(hass, domain_group, config)
-
-    # Enables after dependencies when setting up stage 1 domains
-    async_set_domains_to_be_loaded(hass, stage_1_domains)
+    _LOGGER.info("Setting up stage 0 and 1")
+    for name, domain_group, timeout in stage_0_and_1_domains:
+        if not domain_group:
+            continue
+
+        _LOGGER.info("Setting up %s: %s", name, domain_group)
+        to_be_loaded = domain_group.copy()
+        to_be_loaded.update(
+            dep
+            for domain in domain_group
+            if (integration := integration_cache.get(domain)) is not None
+            for dep in integration.all_dependencies
+        )
+        async_set_domains_to_be_loaded(hass, to_be_loaded)
+        stage_2_domains -= to_be_loaded
 
-    # Start setup
-    if stage_1_domains:
-        _LOGGER.info("Setting up stage 1: %s", stage_1_domains)
         try:
-            async with hass.timeout.async_timeout(
-                STAGE_1_TIMEOUT, cool_down=COOLDOWN_TIME
-            ):
-                await _async_setup_multi_components(hass, stage_1_domains, config)
+            async with hass.timeout.async_timeout(timeout, cool_down=COOLDOWN_TIME):
+                await _async_setup_multi_components(hass, domain_group, config)
```

Comment:
This will cancel the inner coro.  If we end up cancelling the frontend or http deps its effectively a fatal exception because they will never be able to interact with the system

Reply:
If we cancel during normal startup, we'll retry in recovery mode, which means less dependencies and a smaller chance of hanging because some dependency chain never finishes. Maybe in recovery mode we should have no or a much longer timeout?

Also, there's a hardcoded per-domain timeout of 5 minutes: https://github.com/home-assistant/core/blob/954a57e5ac0d3e062cc68970c133dd8246b453e8/homeassistant/setup.py#L421